File: app/src/main/java/org/mifos/mobilebanking/ui/activities/HomeActivity.java
package org.mifos.mobilebanking.ui.activities;


import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.os.Handler;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.api.local.PreferencesHelper;
import org.mifos.mobilebanking.models.client.Client;
import org.mifos.mobilebanking.presenters.UserDetailsPresenter;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.enums.AccountType;
import org.mifos.mobilebanking.ui.enums.ChargeType;
import org.mifos.mobilebanking.ui.fragments.BeneficiaryListFragment;
import org.mifos.mobilebanking.ui.fragments.ClientAccountsFragment;
import org.mifos.mobilebanking.ui.fragments.ClientChargeFragment;
import org.mifos.mobilebanking.ui.fragments.HomeOldFragment;
import org.mifos.mobilebanking.ui.fragments.NotificationFragment;
import org.mifos.mobilebanking.ui.fragments.RecentTransactionsFragment;
import org.mifos.mobilebanking.ui.fragments.ThirdPartyTransferFragment;
import org.mifos.mobilebanking.ui.views.UserDetailsView;
import org.mifos.mobilebanking.utils.CircularImageView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.MaterialDialog;
import org.mifos.mobilebanking.utils.TextDrawable;
import org.mifos.mobilebanking.utils.Toaster;
import org.mifos.mobilebanking.utils.gcm.RegistrationIntentService;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Vishwajeet
 * @since 14/07/2016
 */
public class HomeActivity extends BaseActivity implements UserDetailsView, NavigationView.
        OnNavigationItemSelectedListener, View.OnClickListener {

    @BindView(R.id.navigation_view)
    NavigationView navigationView;

    @BindView(R.id.drawer)
    DrawerLayout drawerLayout;

    @Inject
    PreferencesHelper preferencesHelper;

    @Inject
    UserDetailsPresenter detailsPresenter;

    private TextView tvUsername;
    private CircularImageView ivCircularUserProfilePicture;
    private ImageView ivTextDrawableUserProfilePicture;
    private long clientId;
    private Bitmap userProfileBitmap;
    private Client client;
    private static final int PLAY_SERVICES_RESOLUTION_REQUEST = 9000;
    private boolean isReceiverRegistered;
    private int menuItem;
    boolean doubleBackToExitPressedOnce = false;

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getActivityComponent().inject(this);
        setContentView(R.layout.activity_home);
        ButterKnife.bind(this);
        clientId = preferencesHelper.getClientId();

        setupNavigationBar();
        setToolbarElevation();
        setToolbarTitle(getString(R.string.home));
        replaceFragment(HomeOldFragment.newInstance(), false, R.id.container);

        if (getIntent() != null && getIntent().getBooleanExtra(getString(R.string.notification),
                false)) {
            replaceFragment(NotificationFragment.newInstance(), true, R.id.container);
        }

        if (savedInstanceState == null) {
            detailsPresenter.attachView(this);
            detailsPresenter.getUserDetails();
            detailsPresenter.getUserImage();
            showUserImage(null);
        } else {
            client = savedInstanceState.getParcelable(Constants.USER_DETAILS);
            detailsPresenter.setUserProfile(preferencesHelper.getUserProfileImage());
            showUserDetails(client);
        }

        if (checkPlayServices() && !preferencesHelper.sentTokenToServerState()) {
            // Start IntentService to register this application with GCM.
            Intent intent = new Intent(this, RegistrationIntentService.class);
            startService(intent);
        }

    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(Constants.USER_DETAILS, client);
    }

    @Override
    protected void onPause() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(registerReceiver);
        isReceiverRegistered = false;
        super.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (!isReceiverRegistered) {
            LocalBroadcastManager.getInstance(this).registerReceiver(registerReceiver,
                    new IntentFilter(Constants.REGISTER_ON_SERVER));
            isReceiverRegistered = true;
        }
    }

    /**
     * Called whenever any item is selected in {@link NavigationView}
     *
     * @param item {@link MenuItem} which is selected by the user
     */
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // select which item to open
        setToolbarElevation();
        menuItem = item.getItemId();
        if (menuItem != R.id.item_settings && menuItem != R.id.item_share
                && menuItem != R.id.item_about_us && menuItem != R.id.item_help) {
            // If we have clicked something other than settings or share
            // we can safely clear the back stack as a new fragment will replace
            // the current fragment.
            clearFragmentBackStack();
        }
        switch (item.getItemId()) {
            case R.id.item_home:
                hideToolbarElevation();
                replaceFragment(HomeOldFragment.newInstance(), true, R.id.container);
                break;
            case R.id.item_accounts:
                hideToolbarElevation();
                replaceFragment(ClientAccountsFragment.newInstance(AccountType.SAVINGS),
                        true, R.id.container);
                break;
            case R.id.item_recent_transactions:
                replaceFragment(RecentTransactionsFragment.newInstance(), true, R.id.container);
                break;
            case R.id.item_charges:
                replaceFragment(ClientChargeFragment.newInstance(clientId, ChargeType.CLIENT), true,
                        R.id.container);
                break;
            case R.id.item_third_party_transfer:
                replaceFragment(ThirdPartyTransferFragment.newInstance(), true, R.id.container);
                break;
            case R.id.item_beneficiaries:
                replaceFragment(BeneficiaryListFragment.newInstance(), true, R.id.container);
                break;
            case R.id.item_settings:
                startActivity(new Intent(HomeActivity.this, SettingsActivity.class));
                break;
            case R.id.item_about_us:
                startActivity(new Intent(HomeActivity.this, AboutUsActivity.class));
                break;
            case R.id.item_help:
                startActivity(new Intent(HomeActivity.this, HelpActivity.class));
                break;
            case R.id.item_share:
                Intent i = new Intent(Intent.ACTION_SEND);
                i.setType("text/plain");
                i.putExtra(Intent.EXTRA_TEXT, getString(R.string.playstore_link,
                        getString(R.string.share_msg), getApplication().getPackageName()));
                startActivity(Intent.createChooser(i, getString(R.string.choose)));
                break;
            case R.id.item_logout:
                showLogoutDialog();
                break;
        }

        // close the drawer
        drawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }

    /**
     * Asks users to confirm whether he want to logout or not
     */
    private void showLogoutDialog() {
        new MaterialDialog.Builder().init(HomeActivity.this)
                .setCancelable(false)
                .setMessage(R.string.dialog_logout)
                .setPositiveButton(getString(R.string.logout),
                        new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                preferencesHelper.clear();
                                Intent i = new Intent(HomeActivity.this, LoginActivity.class);
                                i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.
                                        FLAG_ACTIVITY_CLEAR_TASK);
                                startActivity(i);
                                finish();
                            }
                        })
                .setNegativeButton(getString(R.string.cancel),
                        new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                setNavigationViewSelectedItem(R.id.item_home);
                            }
                        })
                .createMaterialDialog()
                .show();
    }

    /**
     * This method is used to set up the navigation drawer for
     * self-service application
     */
    private void setupNavigationBar() {

        navigationView.setNavigationItemSelectedListener(this);

        ActionBarDrawerToggle actionBarDrawerToggle = new ActionBarDrawerToggle(this,
                drawerLayout, toolbar, R.string.open_drawer, R.string.close_drawer) {

            @Override
            public void onDrawerClosed(View drawerView) {
                super.onDrawerClosed(drawerView);
            }

            @Override
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                hideKeyboard(drawerView);
            }
        };
        drawerLayout.addDrawerListener(actionBarDrawerToggle);
        actionBarDrawerToggle.syncState();
        setupHeaderView(navigationView.getHeaderView(0));
        setUpBackStackListener();
    }

    /**
     * Used for initializing values for HeaderView of NavigationView
     *
     * @param headerView Header view of NavigationView
     */
    private void setupHeaderView(View headerView) {
        tvUsername = ButterKnife.findById(headerView, R.id.tv_user_name);
        ivCircularUserProfilePicture = ButterKnife.findById(headerView,
                R.id.iv_circular_user_image);
        ivTextDrawableUserProfilePicture = ButterKnife.findById(headerView, R.id.iv_user_image);

        ivTextDrawableUserProfilePicture.setOnClickListener(this);
        ivCircularUserProfilePicture.setOnClickListener(this);
    }

    /**
     * Shows Client username in HeaderView of NavigationView
     *
     * @param client Contains details about the client
     */
    @Override
    public void showUserDetails(Client client) {
        this.client = client;
        preferencesHelper.setClientName(client.getDisplayName());
        tvUsername.setText(client.getDisplayName());
    }

    /**
     * Displays UserProfile Picture in HeaderView in NavigationView
     *
     * @param bitmap UserProfile Picture
     */
    @Override
    public void showUserImage(final Bitmap bitmap) {
        if (bitmap != null) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    userProfileBitmap = bitmap;
                    ivCircularUserProfilePicture.setImageBitmap(bitmap);
                    ivCircularUserProfilePicture.setVisibility(View.VISIBLE);
                    ivTextDrawableUserProfilePicture.setVisibility(View.GONE);
                }
            });
        } else {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    String userName;
                    if (!preferencesHelper.getClientName().isEmpty()) {
                        userName = preferencesHelper.getClientName();
                    } else {
                        userName = getString(R.string.app_name);
                    }
                    ivCircularUserProfilePicture.setVisibility(View.GONE);
                    ivTextDrawableUserProfilePicture.setVisibility(View.VISIBLE);
                    TextDrawable drawable = TextDrawable.builder()
                            .beginConfig()
                            .toUpperCase()
                            .endConfig()
                            .buildRound(userName.substring(0, 1),
                                    ContextCompat.getColor(
                                            HomeActivity.this, R.color.primary_dark));
                    ivTextDrawableUserProfilePicture.setImageDrawable(drawable);
                }
            });
        }
    }

    @Override
    public void showProgress() {
        //empty, no need to show/hide progress in headerview
    }

    @Override
    public void hideProgress() {
        //empty
    }

    /**
     * It is called whenever any error occurs while executing a request
     *
     * @param message contains information about error occurred
     */
    @Override
    public void showError(String message) {
        showToast(message, Toast.LENGTH_SHORT);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        detailsPresenter.detachView();
    }

    /**
     * Handling back press
     */
    @Override
    public void onBackPressed() {

        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START);
            return;
        }

        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.container);
        if (fragment instanceof HomeOldFragment) {
            if (doubleBackToExitPressedOnce && stackCount() == 0) {
                HomeActivity.this.finish();
                return;
            }
            this.doubleBackToExitPressedOnce = true;
            Toaster.show(findViewById(android.R.id.content), getString(R.string.exit_message));
            new Handler().postDelayed(new Runnable() {

                @Override
                public void run() {
                    doubleBackToExitPressedOnce = false;
                }
            }, 2000);
        }

        if (stackCount() != 0) {
            super.onBackPressed();
        }
    }

    private void setUpBackStackListener() {
        getSupportFragmentManager().
                addOnBackStackChangedListener(new FragmentManager.OnBackStackChangedListener() {
                    @Override
                    public void onBackStackChanged() {
                        Fragment fragment = getSupportFragmentManager().
                                findFragmentById(R.id.container);
                        setToolbarElevation();
                        if (fragment instanceof HomeOldFragment) {
                            setNavigationViewSelectedItem(R.id.item_home);
                        } else if (fragment instanceof ClientAccountsFragment) {
                            hideToolbarElevation();
                            setNavigationViewSelectedItem(R.id.item_accounts);
                        } else if (fragment instanceof RecentTransactionsFragment) {
                            setNavigationViewSelectedItem(R.id.item_recent_transactions);
                        } else if (fragment instanceof ClientChargeFragment) {
                            setNavigationViewSelectedItem(R.id.item_charges);
                        } else if (fragment instanceof ThirdPartyTransferFragment) {
                            setNavigationViewSelectedItem(R.id.item_third_party_transfer);
                        } else if (fragment instanceof BeneficiaryListFragment) {
                            setNavigationViewSelectedItem(R.id.item_beneficiaries);
                        }
                    }
                });
    }

    public void setNavigationViewSelectedItem(int id) {
        navigationView.setCheckedItem(id);
    }

    @Override
    public void onClick(View v) {
        // Click Header to view full profile of User
        startActivity(new Intent(HomeActivity.this, UserProfileActivity.class));
        drawerLayout.closeDrawer(GravityCompat.START);
    }

    /**
     * Check the device to make sure it has the Google Play Services APK. If
     * it doesn't, display a dialog that allows users to download the APK from
     * the Google Play Store or enable it in the device's system settings.
     */
    private boolean checkPlayServices() {
        GoogleApiAvailability apiAvailability = GoogleApiAvailability.getInstance();
        int resultCode = apiAvailability.isGooglePlayServicesAvailable(this);
        if (resultCode != ConnectionResult.SUCCESS) {
            if (apiAvailability.isUserResolvableError(resultCode)) {
                apiAvailability.getErrorDialog(this, resultCode, PLAY_SERVICES_RESOLUTION_REQUEST)
                        .show();
            } else {
                Log.i(HomeActivity.class.getName(), "This device is not supported.");
                finish();
            }
            return false;
        }
        return true;
    }

    private BroadcastReceiver registerReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String token = intent.getStringExtra(Constants.TOKEN);
            detailsPresenter.registerNotification(token);
        }
    };

    public int getCheckedItem() {
        return menuItem;
    }

    public void hideKeyboard(View view) {
        InputMethodManager inputManager = (InputMethodManager) this
                .getSystemService(Context.INPUT_METHOD_SERVICE);
        inputManager.hideSoftInputFromWindow(view.getWindowToken(), InputMethodManager
                .RESULT_UNCHANGED_SHOWN);
    }

}

File: app/src/main/java/org/mifos/mobile/ui/activities/HomeActivity.kt
package org.mifos.mobile.ui.activities

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.core.view.GravityCompat
import androidx.lifecycle.lifecycleScope
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.imageview.ShapeableImageView
import com.google.android.material.navigation.NavigationView
import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobile.R
import org.mifos.mobile.api.local.PreferencesHelper
import org.mifos.mobile.databinding.ActivityHomeBinding
import org.mifos.mobile.databinding.NavDrawerHeaderBinding
import org.mifos.mobile.models.client.Client
import org.mifos.mobile.ui.about.AboutUsActivity
import org.mifos.mobile.ui.activities.base.BaseActivity
import org.mifos.mobile.ui.enums.AccountType
import org.mifos.mobile.ui.enums.ChargeType
import org.mifos.mobile.ui.fragments.*
import org.mifos.mobile.ui.getThemeAttributeColor
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.utils.TextDrawable
import org.mifos.mobile.utils.Toaster
import org.mifos.mobile.utils.UserDetailUiState
import org.mifos.mobile.utils.fcm.RegistrationIntentService
import org.mifos.mobile.viewModels.UserDetailViewModel
import javax.inject.Inject

/**
 * @author Vishwajeet
 * @since 14/07/2016
 */
@AndroidEntryPoint
class HomeActivity :
    BaseActivity(),
    NavigationView.OnNavigationItemSelectedListener {

    private lateinit var binding: ActivityHomeBinding
    private lateinit var navHeaderBinding: NavDrawerHeaderBinding

    @JvmField
    @Inject
    var preferencesHelper: PreferencesHelper? = null

    private val viewModel: UserDetailViewModel by viewModels()

    private var tvUsername: TextView? = null
    private var drawerUserImage: ShapeableImageView? = null
    private var clientId: Long? = 0
    private var userProfileBitmap: Bitmap? = null
    private var client: Client? = null
    private var isReceiverRegistered = false
    var checkedItem = 0
        private set
    var doubleBackToExitPressedOnce = false
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityHomeBinding.inflate(layoutInflater)
        setContentView(binding.root)
        clientId = preferencesHelper?.clientId
        setupNavigationBar()
        setToolbarElevation()
        setToolbarTitle(getString(R.string.home))
        replaceFragment(HomeOldFragment.newInstance(), false, R.id.container)
        if (intent != null && intent.getBooleanExtra(
                getString(R.string.notification),
                false,
            )
        ) {
            replaceFragment(NotificationFragment.newInstance(), true, R.id.container)
        }
        if (savedInstanceState == null) {
            viewModel.userDetails
            viewModel.userImage
            showUserImage(null)
        } else {
            client = savedInstanceState.getParcelable(Constants.USER_DETAILS)
            viewModel.setUserProfile(preferencesHelper?.userProfileImage)
            showUserDetails(client)
        }
        if (checkPlayServices() && preferencesHelper?.sentTokenToServerState() == false) {
            // Start IntentService to register this application with GCM.
            val intent = Intent(this, RegistrationIntentService::class.java)
            startService(intent)
        }

        lifecycleScope.launchWhenStarted {
            viewModel.userDetailUiState.collect {
                when (it) {
                    is UserDetailUiState.Loading -> showProgress()
                    is UserDetailUiState.ShowUserDetails -> {
                        hideProgress()
                        showUserDetails(it.client)
                    }
                    is UserDetailUiState.ShowUserImage -> {
                        hideProgress()
                        showUserImage(it.image)
                    }
                    is UserDetailUiState.ShowError -> {
                        hideProgress()
                        showError(getString(it.message))
                    }
                }
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(Constants.USER_DETAILS, client)
    }

    override fun onPause() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(registerReceiver)
        isReceiverRegistered = false
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        if (!isReceiverRegistered) {
            LocalBroadcastManager.getInstance(this).registerReceiver(
                registerReceiver,
                IntentFilter(Constants.REGISTER_ON_SERVER),
            )
            isReceiverRegistered = true
        }
    }

    /**
     * Called whenever any item is selected in [NavigationView]
     *
     * @param item [MenuItem] which is selected by the user
     */
    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // select which item to open
        setToolbarElevation()
        checkedItem = item.itemId
        if (checkedItem != R.id.item_settings && checkedItem != R.id.item_share && checkedItem != R.id.item_about_us && checkedItem != R.id.item_help) {
            // If we have clicked something other than settings or share
            // we can safely clear the back stack as a new fragment will replace
            // the current fragment.
            clearFragmentBackStack()
        }
        when (item.itemId) {
            R.id.item_home -> {
                hideToolbarElevation()
                replaceFragment(HomeOldFragment.newInstance(), true, R.id.container)
            }

            R.id.item_accounts -> {
                hideToolbarElevation()
                replaceFragment(
                    ClientAccountsFragment.newInstance(AccountType.SAVINGS),
                    true,
                    R.id.container,
                )
            }

            R.id.item_recent_transactions -> replaceFragment(
                RecentTransactionsFragment.newInstance(),
                true,
                R.id.container,
            )

            R.id.item_charges -> replaceFragment(
                ClientChargeFragment.newInstance(clientId, ChargeType.CLIENT),
                true,
                R.id.container,
            )

            R.id.item_third_party_transfer -> replaceFragment(
                ThirdPartyTransferFragment.newInstance(),
                true,
                R.id.container,
            )

            R.id.item_beneficiaries -> replaceFragment(
                BeneficiaryListFragment.newInstance(),
                true,
                R.id.container,
            )

            R.id.item_settings -> startActivity(
                Intent(
                    this@HomeActivity,
                    SettingsActivity::class.java,
                ),
            )

            R.id.item_about_us -> startActivity(
                Intent(
                    this@HomeActivity,
                    AboutUsActivity::class.java,
                ),
            )

            R.id.item_help -> startActivity(Intent(this@HomeActivity, HelpActivity::class.java))
            R.id.item_app_info -> {
                val intent = Intent(android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                intent.data = Uri.parse("package:$packageName")
                startActivity(intent)
            }

            R.id.item_share -> {
                val i = Intent(Intent.ACTION_SEND)
                i.type = "text/plain"
                i.putExtra(
                    Intent.EXTRA_TEXT,
                    getString(
                        R.string.playstore_link,
                        getString(R.string.share_msg),
                        application.packageName,
                    ),
                )
                startActivity(Intent.createChooser(i, getString(R.string.choose)))
            }

            R.id.item_logout -> showLogoutDialog()
        }

        // close the drawer
        binding.drawer.closeDrawer(GravityCompat.START)
        return true
    }

    /**
     * Asks users to confirm whether he want to logout or not
     */
    private fun showLogoutDialog() {
        MaterialAlertDialogBuilder(this, R.style.RedDialog)
            .setTitle(R.string.dialog_logout)
            .setIcon(R.drawable.ic_logout)
            .setPositiveButton(getString(R.string.logout)) { _, _ ->
                preferencesHelper?.clear()
                val i = Intent(this@HomeActivity, LoginActivity::class.java)
                i.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                startActivity(i)
                finish()
                Toast.makeText(this, R.string.logged_out_successfully, Toast.LENGTH_SHORT).show()
            }
            .setNegativeButton(getString(R.string.cancel)) { dialogInterface, _ -> dialogInterface.dismiss() }
            .create()
            .show()
    }

    /**
     * This method is used to set up the navigation drawer for
     * self-service application
     */
    private fun setupNavigationBar() {
        binding.navigationView.setNavigationItemSelectedListener(this)
        val actionBarDrawerToggle: ActionBarDrawerToggle = object : ActionBarDrawerToggle(
            this,
            binding.drawer,
            toolbar,
            R.string.open_drawer,
            R.string.close_drawer,
        ) {

            override fun onDrawerOpened(drawerView: View) {
                super.onDrawerOpened(drawerView)
                hideKeyboard(drawerView)
            }
        }
        binding.drawer.addDrawerListener(actionBarDrawerToggle)
        actionBarDrawerToggle.syncState()
        binding.navigationView.getHeaderView(0)?.let { setupHeaderView(it) }
        setUpBackStackListener()
    }

    /**
     * Used for initializing values for HeaderView of NavigationView
     *
     * @param headerView Header view of NavigationView
     */
    private fun setupHeaderView(headerView: View) {
        navHeaderBinding = NavDrawerHeaderBinding.bind(headerView)
        tvUsername = navHeaderBinding.tvUserName
        drawerUserImage = navHeaderBinding.userImageRound
        drawerUserImage?.setOnClickListener {
            startActivity(Intent(this, UserProfileActivity::class.java))
            binding.drawer.closeDrawer(GravityCompat.START)
        }
    }

    /**
     * Shows Client username in HeaderView of NavigationView
     *
     * @param client Contains details about the client
     */
    fun showUserDetails(client: Client?) {
        this.client = client
        preferencesHelper?.clientName = client?.displayName
        tvUsername?.text = client?.displayName
    }

    /**
     * Displays UserProfile Picture in HeaderView in NavigationView
     *
     * @param bitmap UserProfile Picture
     */
    fun showUserImage(bitmap: Bitmap?) {
        if (bitmap != null) {
            runOnUiThread {
                userProfileBitmap = bitmap
                drawerUserImage?.setImageBitmap(bitmap)
            }
        } else {
            runOnUiThread {
                val userName: String? = if (preferencesHelper?.clientName?.isEmpty() == false) {
                    preferencesHelper?.clientName
                } else {
                    getString(R.string.app_name)
                }
                val drawable = TextDrawable.builder()
                    .beginConfig()
                    .toUpperCase()
                    .endConfig()
                    .buildRound(
                        userName?.substring(0, 1),
                        getThemeAttributeColor(R.attr.colorPrimaryVariant),
                    )
                drawerUserImage?.setImageDrawable(drawable)
            }
        }
    }

    fun showProgress() {
        // empty, no need to show/hide progress in headerview
    }

    fun hideProgress() {
        // empty
    }

    /**
     * It is called whenever any error occurs while executing a request
     *
     * @param message contains information about error occurred
     */
    fun showError(message: String?) {
        showToast(message, Toast.LENGTH_SHORT)
    }

    /**
     * Handling back press
     */
    override fun onBackPressed() {
        if (binding.drawer.isDrawerOpen(GravityCompat.START)) {
            binding.drawer.closeDrawer(GravityCompat.START)
            return
        }
        val fragment = supportFragmentManager.findFragmentById(R.id.container)
        if (fragment is HomeOldFragment) {
            if (doubleBackToExitPressedOnce && stackCount() == 0) {
                finish()
                return
            }
            doubleBackToExitPressedOnce = true
            Toaster.show(findViewById(android.R.id.content), getString(R.string.exit_message))
            Handler().postDelayed({ doubleBackToExitPressedOnce = false }, 2000)
        } else if (fragment is TransferProcessFragment) {
            fragment.cancelTransferProcess()
        }

        if (stackCount() != 0) {
            super.onBackPressed()
        }
    }

    private fun setUpBackStackListener() {
        supportFragmentManager.addOnBackStackChangedListener {
            val fragment = supportFragmentManager.findFragmentById(R.id.container)
            setToolbarElevation()
            when (fragment) {
                is HomeOldFragment -> {
                    setNavigationViewSelectedItem(R.id.item_home)
                }

                is ClientAccountsFragment -> {
                    hideToolbarElevation()
                    setNavigationViewSelectedItem(R.id.item_accounts)
                }

                is RecentTransactionsFragment -> {
                    setNavigationViewSelectedItem(R.id.item_recent_transactions)
                }

                is ClientChargeFragment -> {
                    setNavigationViewSelectedItem(R.id.item_charges)
                }

                is ThirdPartyTransferFragment -> {
                    setNavigationViewSelectedItem(R.id.item_third_party_transfer)
                }

                is BeneficiaryListFragment -> {
                    setNavigationViewSelectedItem(R.id.item_beneficiaries)
                }
            }
        }
    }

    fun setNavigationViewSelectedItem(id: Int) {
        binding.navigationView.setCheckedItem(id)
    }

    /**
     * Check the device to make sure it has the Google Play Services APK. If
     * it doesn't, display a dialog that allows users to download the APK from
     * the Google Play Store or enable it in the device's system settings.
     */
    private fun checkPlayServices(): Boolean {
        val apiAvailability = GoogleApiAvailability.getInstance()
        val resultCode = apiAvailability.isGooglePlayServicesAvailable(this)
        if (resultCode != ConnectionResult.SUCCESS) {
            if (apiAvailability.isUserResolvableError(resultCode)) {
                apiAvailability.getErrorDialog(this, resultCode, PLAY_SERVICES_RESOLUTION_REQUEST)
                    ?.show()
            } else {
                Log.i(HomeActivity::class.java.name, "This device is not supported.")
                finish()
            }
            return false
        }
        return true
    }

    private val registerReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val token = intent.getStringExtra(Constants.TOKEN)
            token?.let { viewModel.registerNotification(it) }
        }
    }

    fun hideKeyboard(view: View) {
        val inputManager = this
            .getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        inputManager.hideSoftInputFromWindow(
            view.windowToken,
            InputMethodManager.RESULT_UNCHANGED_SHOWN,
        )
    }

    companion object {
        private const val PLAY_SERVICES_RESOLUTION_REQUEST = 9000
    }
}
